### DISCUSSION TITLE
labwc 0.8.2 : spacefm stopped starting

### DISCUSSION THREAD

**Discussion by sfs-pra** [labwc/labwc#2447] [Q&A] [answered]
`segmentation fault`
labwc 0.8.1 everything was working. Didn't change anything else
The author stopped support https://aur.archlinux.org/packages/spacefm. I don't know where to turn anymore

**Comment by tokyo4j** [comment]
Looks like spacefm segfaults with labwc starting from 64bb57af, but I have no idea why it causes issues with spacefm.

**Reply by Consolatis** [reply]
I assume the segfault is the consequence of the wayland connection dropping. `WAYLAND_DEBUG=1` might shed some light on this.

**Comment by sfs-pra** [comment]
Removing https://github.com/labwc/labwc/commit/64bb57afa22b4ecded17a271392964b35e8d780f solves the problem. What do I have to lose this way?
Attached WAYLAND_DEBUG=1 spacefm 
[sp.log](https://github.com/user-attachments/files/18232280/sp.log)

**Reply by Consolatis** [reply]
GTK does bind the importer and exporter interfaces of the xdg-foreign-v1 protocol but the log doesn't seem to indicate a reason for the crash. This is something that must likely be debugged via the crash dump or running spacefm via gdb. As its a C++ application I don't have the resources for that though.

Another option that could potentially work around the issue is using `GDK_BACKEND=x11 spacefm` to force it to run via xwayland. Note that drag-and-drop from spacefm to wayland applications will then be broken though.

> What do I have to lose this way?

Not much, I assume mostly a file-open/save dialog via xdg-portal.

**Reply by tokyo4j** [reply]
I got the backtrace:

```
(gdb) bt
#0  0x0000000000000000 in ??? ()
#1  0x00007ffff7192d9c in g_hash_table_lookup () at /usr/lib/libglib-2.0.so.0
#2  0x00007ffff7f3ad22 in gdk_x11_window_lookup_for_display () at /usr/lib/libgdk-3.so.0
#3  0x000055555563d4da in get_desktop_index (win=0x0) at main-window.c:4179
#4  0x000055555563d766 in fm_main_window_get_on_current_desktop () at main-window.c:4225
#5  0x00005555556306d9 in handle_parsed_commandline_args () at main.c:1143
#6  0x000055555563164a in main (argc=1, argv=0x7fffffffe808) at main.c:1526
```

And here's the relevant part:

https://github.com/IgnorantGuru/spacefm/blob/e6f291858067e73db44fb57c90e4efb97b088ac8/src/main-window.c#L4175-L4181

Haven't looked in depth yet though.

**Reply by Consolatis** [reply]
Based on the `gdk_x11_window_lookup_for_display()` function call this looks like its a general wayland issue with spacefm expecting it to be running via X11 rather than wayland in some specific code path triggered by the availability of the xdg-foreign-v1 interfaces.

**Reply by Consolatis** [reply]
This seems to be relevant: https://github.com/IgnorantGuru/spacefm/pull/801
@sfs-pra if you run spacefm without the foreign protocols, does it show up as `[xwayland]` or `[xdg]` in `A-Tab` ? Maybe what happens is that GTK uses the availability of the xdg-foreign-v1 protocol to decide on using the wayland or x11 backend.

**Reply by sfs-pra** [reply]
https://github.com/puppylinux-woof-CE/woof-CE/blob/testing/woof-code/rootfs-petbuilds/spacefm/wayland.patch Did not solve the problem

**Comment by sfs-pra** [comment]
Without https://github.com/labwc/labwc/commit/64bb57afa22b4ecded17a271392964b35e8d780f - spacefm starts and a-tab is visible

**Reply by Consolatis** [reply]
The question is if it runs via xwayland or wayland native. You can also check if it shows up in `wmctrl -l`.

**Reply by sfs-pra** [reply]
```
$ pgrep -l spacefm && wmctrl -l
505393 spacefm
```

**Reply by Consolatis** [reply]
Yeah, that looks like spacefm uses the X11 backend when GTK does not find xdg-foreign-v1 and tries to use the wayland backend otherwise. spacefm however does not support wayland and instead uses X11 specific calls which fail when GTK is running on the wayland backend. The solution is thus either of
- `GDK_BACKEND=x11 spacefm` (dnd from spacefm to wayland applications will not work)
- getting wayland support into spacefm (e.g. making it not use X11 calls when running on wayland)

**Comment by sfs-pra** [comment, marked as answer]
Double-checked
https://github.com/puppylinux-woof-CE/woof-CE/blob/testing/woof-code/rootfs-petbuilds/spacefm/wayland.patch solves the problem
[PKGBUILD.zip](https://github.com/user-attachments/files/18246731/PKGBUILD.zip)
